### ISSUE TITLE
Crash in ConcurrentMergeScheduler

### ISSUE THREAD

**Issue by vadz** [luceneplusplus/LucenePlusPlus#86]
**TL;DR**: If you have mysterious crashes in `LuceneThread::start()`, switch to using `SerialMergeScheduler`.

----
There is a nasty race condition in `ConcurrentMergeScheduler`: its `merge()` method creates a new `LuceneThread` object and takes care to add it to `mergeThreads`, which ensures that this object is not deleted immediately, before its `runThread()` is executed in the new thread.

However there is absolutely nothing done to ensure that this thread finishes, or even starts, by the time `ConcurrentMergeScheduler` itself is destroyed. So it can perfectly well happen (and, as I can assure you after spending half a day debugging this crash which I couldn't reproduce on my machine but which happened reliably on the user's one, *does* happen) that `IndexWriter`, containing `ConcurrentMergeScheduler`, is destroyed before the thread starts running and by the time it does, `thread` argument of `runThread()` is invalid as the object had been destroyed.

I am not sure what is the right fix for this, the thread class looks very dangerous to me and personally I'd rewrite it in some safer way ensuring that things like this can't happen. For now we've just switched to using `SerialMergeScheduler` though, so I'm mostly writing this to help others running into this problem.

**Comment by alanw** [comment]
Thanks for reporting this vadz. 

Regarding the thread class, I completely agree - I've never been happy with LuceneThread. The reason for it's apparent complexity is due to it's cross-platform nature (threads have slightly different characteristics on Windows compared to Unix). I'll give this some thought.

In the meantime, could I ask that you check whether IndexWriter::close() is called before it's destroyed? This is supposed to wait for all merge threads to finish, but it sounds like there's a subtle bug lurking which causes this problem.

If you continue to experience the same problem, then I'll perhaps consider switching the default MergeScheduler to SerialMergeScheduler.

**Comment by vadz** [comment]
Sorry, I have real trouble reproducing this -- not the bug itself, but actually executing the code that triggers it. I've just spent half an hour trying to do it unsuccessfully and I don't think I can afford spending more time on this, especially as I have a simple workaround for the problem.

To be precise, the trouble is that `ConcurrentMergeScheduler::merge()` returns without doing anything interesting because `writer->getNextMerge()` returns null and so the code calling `getMergeThread()` and actually creating the thread is never executed. Of course, I know that it does happen sometimes (because I have dumps of crashes to prove it...), but I have no idea how to trigger it intentionally. If you can advise me how to do it, I could retry it again but otherwise I'm afraid I'll have to stop here.
